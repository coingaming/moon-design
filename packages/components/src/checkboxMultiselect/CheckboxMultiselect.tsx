import { rem } from '@heathmont/moon-utils';
import React from 'react';
import styled from 'styled-components';

import Checkbox from '../checkbox/Checkbox';

type Option = {
  label: string | JSX.Element;
  value: any;
  [key: string]: any;
};

type CheckboxMultiselectProps = {
  value?: any[];
  options: Option[];
  onChange: (value: any[]) => void;
  maxHeight?: string | number;
};

const List = styled.ul(() => ({
  overflow: 'auto',
  listStyle: 'none',
  padding: 0,
  margin: `0 -${rem(12)}`,
}));

const Option = styled.label(({ theme }) => ({
  display: 'flex',
  alignItems: 'center',
  fontSize: rem(14),
  lineHeight: rem(20),
  padding: `${rem(10)} ${rem(12)}`,
  textDecoration: 'none',
  outline: 'none',
  border: 'none',
  textAlign: 'left',
  width: '100%',
  cursor: 'pointer',
  color: theme.color.bulma[100],
  backgroundColor: 'transparent',
  'span::before': {
    display: 'none',
  },
  '&:hover, &:focus': {
    backgroundColor: theme.color.goku[100],
  },
}));

const CheckboxMultiselect: React.FC<CheckboxMultiselectProps> = ({
  options,
  value = [],
  onChange,
  maxHeight,
}) => {
  const onCheckboxChange = (optionValue: any) => {
    if (value.includes(optionValue)) {
      onChange(value.filter(currentValue => currentValue !== optionValue));
    } else {
      onChange([...value, optionValue]);
    }
  };

  return (
    <List style={{ maxHeight }}>
      {options.map(option => (
        <li key={option.value}>
          <Option>
            <Checkbox
              id={option.id}
              checked={value.some(
                currentValue => currentValue === option.value
              )}
              onChange={() => onCheckboxChange(option.value)}
              label={option.label}
            />
            {/* {option.label} */}
          </Option>
        </li>
      ))}
    </List>
  );
};

CheckboxMultiselect.defaultProps = {
  maxHeight: 250,
};

export default CheckboxMultiselect;
