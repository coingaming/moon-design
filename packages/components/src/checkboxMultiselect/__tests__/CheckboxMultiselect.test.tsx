import React from 'react';
import { act, create, ReactTestRenderer } from 'react-test-renderer';
import { sportsbetDark, ThemeProvider } from '@heathmont/moon-themes';
import 'jest-styled-components';

import CheckboxMultiselect from '../CheckboxMultiselect';

const renderWithTheme = (component: JSX.Element) => (
  <ThemeProvider theme={sportsbetDark}>{component}</ThemeProvider>
);

describe('CheckboxMultiselect', () => {
  test('renders by default', () => {
    const checkboxMultiselect = create(
      renderWithTheme(
        <CheckboxMultiselect
          onChange={() => {}}
          clearSelectedLabel={<span>Clear selected items</span>}
          value={[]}
          maxHeight={250}
          options={[
            { label: 'Game 1', id: 1 },
            { label: 'Game 2', id: 2 },
            { label: 'Game 3', id: 3 },
          ]}
        />
      )
    );

    expect(checkboxMultiselect).toMatchSnapshot();
  });

  describe('when renders with onChange handler', () => {
    const onChangeMock = jest.fn();
    let checkboxMultiselect: ReactTestRenderer;

    beforeAll(() => {
      checkboxMultiselect = create(
        renderWithTheme(
          <CheckboxMultiselect
            onChange={onChangeMock}
            clearSelectedLabel={<span>Clear selected items</span>}
            value={[]}
            maxHeight={250}
            options={[
              { label: 'Game 1' },
              { label: 'Game 2' },
              { label: 'Game 3' },
            ]}
          />
        )
      );
    });

    describe('and then clicking on the first option', () => {
      beforeAll(() => {
        const input = checkboxMultiselect.root.findAllByType('input')[0];
        act(() => input.props.onChange({ target: { checked: true } }));
      });

      test('calls onChange handler with selected option', () => {
        expect(onChangeMock).toHaveBeenCalledWith([{ label: 'Game 1' }]);
      });
    });
  });

  describe('when renders with onChange handler and pre-selected value', () => {
    const onChangeMock = jest.fn();
    let checkboxMultiselect: ReactTestRenderer;

    beforeAll(() => {
      checkboxMultiselect = create(
        renderWithTheme(
          <CheckboxMultiselect
            onChange={onChangeMock}
            clearSelectedLabel={<span>Clear selected items</span>}
            value={[{ label: 'Game 1' }]}
            maxHeight={250}
            options={[
              { label: 'Game 1' },
              { label: 'Game 2' },
              { label: 'Game 3' },
            ]}
          />
        )
      );
    });

    describe('and then clicking on "clear selection" link', () => {
      beforeAll(() => {
        const clearSelection = checkboxMultiselect.root.findByType('a');
        act(() => clearSelection.props.onClick({ preventDefault: () => {} }));
      });

      test('calls onChange handler with empty array', () => {
        expect(onChangeMock).toHaveBeenCalledWith([]);
      });
    });
  });
});
