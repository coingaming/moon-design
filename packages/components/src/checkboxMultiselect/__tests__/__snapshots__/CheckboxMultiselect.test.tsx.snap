// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxMultiselect renders by default 1`] = `
.c2 {
  display: inline-block;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 1rem;
}

.c5 {
  display: inline-block;
  margin-left: 1.75rem;
  color: #8697A2;
}

.c5::before,
.c5::after {
  content: "";
  position: absolute;
  width: 1rem;
  height: 1rem;
  top: 0.0625rem;
  left: 0;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  -webkit-transition-timing-function: ease;
  transition-timing-function: ease;
}

.c5::before {
  background-color: #4C5159;
  border-radius: 50%;
  opacity: 0;
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
  -webkit-transition-property: background-color,-webkit-transform,opacity;
  -webkit-transition-property: background-color,transform,opacity;
  transition-property: background-color,transform,opacity;
  will-change: transform,opacity;
}

.c5::after {
  border: 1px solid #8697A2;
  border-radius: 0.25rem;
  background-color: transparent;
  -webkit-transition-property: border-color;
  transition-property: border-color;
}

.c3 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c3:hover:enabled + span::before,
.c3:focus:enabled + span::before {
  opacity: 0.5;
  -webkit-transform: scale(2);
  -ms-transform: scale(2);
  transform: scale(2);
}

.c3:hover:enabled:checked + span::before,
.c3:focus:enabled:checked + span::before {
  opacity: 0.1;
  background-color: #19E170;
}

.c3:hover:enabled:not(:checked) + span::after,
.c3:focus:enabled:not(:checked) + span::after {
  border-color: #8697A2;
}

.c3:checked + span::after {
  background-image: url('data:image/svg+xml;utf8, %3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%22100%25%22%20height%3D%22auto%22%20fill%3D%22none%22%20viewBox%3D%220%200%2020%2014%22%3E%3Cpath%20fill%3D%22%230CD463%22%20fill-rule%3D%22evenodd%22%20d%3D%22M18%200c-.56%200-1.06.22-1.42.58L8%209.18l-4.58-4.6C3.06%204.22%202.56%204%202%204%20.9%204%200%204.9%200%206c0%20.56.22%201.06.58%201.42l6%206c.36.36.86.58%201.42.58.56%200%201.06-.22%201.42-.58l10-10c.36-.36.58-.86.58-1.42%200-1.1-.9-2-2-2z%22%20clip-rule%3D%22evenodd%22%3E%3C%2Fpath%3E%3C%2Fsvg%3E');
  background-repeat: no-repeat;
  background-position: center;
  background-size: 0.625rem;
  border-color: #0CD463;
}

.c3[disabled] + span {
  color: #353C44;
  cursor: not-allowed;
}

.c3[disabled] + span::after {
  border-color: #353C44;
}

.c0 {
  list-style: none;
  max-height: 250px;
  overflow: scroll;
  padding-left: 0.5rem;
}

.c1 {
  margin: 1.5rem 0;
}

.c1:first-child {
  margin-top: 0.75rem;
}

.c1:last-child {
  margin-bottom: 0.75rem;
}

.c4 + span::after {
  border-radius: 0.125rem;
}

.c4:checked + span::after {
  background-color: #0CD463;
  background-image: url('data:image/svg+xml;utf8, %3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%22100%25%22%20height%3D%22auto%22%20fill%3D%22none%22%20viewBox%3D%220%200%2020%2014%22%3E%3Cpath%20fill%3D%22%23FFFFFF%22%20fill-rule%3D%22evenodd%22%20d%3D%22M18%200c-.56%200-1.06.22-1.42.58L8%209.18l-4.58-4.6C3.06%204.22%202.56%204%202%204%20.9%204%200%204.9%200%206c0%20.56.22%201.06.58%201.42l6%206c.36.36.86.58%201.42.58.56%200%201.06-.22%201.42-.58l10-10c.36-.36.58-.86.58-1.42%200-1.1-.9-2-2-2z%22%20clip-rule%3D%22evenodd%22%3E%3C%2Fpath%3E%3C%2Fsvg%3E');
}

<div>
  <ul
    className="c0"
  >
    <li
      className="c1"
    >
      <label
        className="c2"
        htmlFor={1}
      >
        <input
          checked={false}
          className="c3 c4"
          disabled={false}
          id={1}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="c5"
        >
          Game 1
        </span>
      </label>
    </li>
    <li
      className="c1"
    >
      <label
        className="c2"
        htmlFor={2}
      >
        <input
          checked={false}
          className="c3 c4"
          disabled={false}
          id={2}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="c5"
        >
          Game 2
        </span>
      </label>
    </li>
    <li
      className="c1"
    >
      <label
        className="c2"
        htmlFor={3}
      >
        <input
          checked={false}
          className="c3 c4"
          disabled={false}
          id={3}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="c5"
        >
          Game 3
        </span>
      </label>
    </li>
  </ul>
</div>
`;
